### PR TITLE
feat: implement screening service lookup

### PR DESCRIPTION
**Pull request by andrewmorris43**
## Description

Adds a database lookup Into `receiveCaasFile` to get the ScreeningId and ScreeningName using the screening acronym from the filename.

This then replaces the hardcoded ScreeningId used in some participant management functions, and in the exception handler function.

I have tested this end to end locally with adding, updating and removing participants.

## Context
[DTOSS-3449](https://nhsd-jira.digital.nhs.uk/browse/DTOSS-3449)

## Type of changes

- [ ] Refactoring (non-breaking change)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would change existing functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [x] I am familiar with the [contributing guidelines](../docs/CONTRIBUTING.md)
- [x] I have followed the code style of the project
- [x] I have added tests to cover my changes
- [ ] I have updated the documentation accordingly
- [ ] This PR is a result of pair or mob programming

---

## Sensitive Information Declaration

To ensure the utmost confidentiality and protect your and others privacy, we kindly ask you to NOT including [PII (Personal Identifiable Information) / PID (Personal Identifiable Data)](https://digital.nhs.uk/data-and-information/keeping-data-safe-and-benefitting-the-public) or any other sensitive data in this PR (Pull Request) and the codebase changes. We will remove any PR that do contain any sensitive information. We really appreciate your cooperation in this matter.

- [x] I confirm that neither PII/PID nor sensitive data are included in this PR and the codebase changes.

### CODE REVIEW

File: application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/Program.cs [hunk 1]
```diff
@@ -2,14 +2,26 @@
...
+        services.AddTransient<IDbConnection>(provider =>
+        {
+            var providerFactory = DbProviderFactories.GetFactory("System.Data.SqlClient");
+            var conn = providerFactory.CreateConnection();
+            return conn;
+        });
         services.AddApplicationInsightsTelemetryWorkerService();
         services.ConfigureFunctionsApplicationInsights();
         services.AddScoped<ICallFunction, CallFunction>();
+        services.AddSingleton<IScreeningServiceData, ScreeningServiceData>();
```
Comment (MWClayson-NHS) [inline application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/Program.cs:24]: Can you change the IDbConnection to be a singleton? 
as a general rule of thumb a singleton should not depend upon a object with a shorter lifetime as there is a chance that the shorter lifetime object will get GC'ed and you'll get errors.

File: application/CohortManager/src/Functions/Shared/Data/Database/ScreeningServiceData.cs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    public ScreeningServiceData(IDbConnection dbConnection, ILogger<ScreeningServiceData> logger)
+    {
+        _dbConnection = dbConnection;
+        _logger = logger;
+        _connectionString = Environment.GetEnvironmentVariable("DtOsDatabaseConnectionString") ?? string.Empty;
+    }
+
+    public ScreeningService GetScreeningServiceByAcronym(string screeningAcronym)
+    {
+        var SQL = "SELECT * " +
```
Comment (MWClayson-NHS) [inline application/CohortManager/src/Functions/Shared/Data/Database/ScreeningServiceData.cs:22]: Please explicitly define the columns you want returning.

File: application/CohortManager/src/Functions/Shared/Data/Database/ScreeningServiceData.cs [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    private ScreeningService GetScreeningService(IDbCommand command)
+    {
+        return ExecuteQuery(command, reader =>
+        {
+            var screeningService = new ScreeningService();
+            while (reader.Read())
+            {
+                screeningService.ScreeningId = reader["SCREENING_ID"] == DBNull.Value ? null : reader["SCREENING_ID"].ToString();
```
Comment (MWClayson-NHS) [inline application/CohortManager/src/Functions/Shared/Data/Database/ScreeningServiceData.cs:44]: there should be DBHelper classes to do this logic for you around checking if a value is null

**Review by MWClayson-NHS — COMMENTED** [review]
Please take a look at my comments, Most of this looks good.
Have we tested this end to end?
I assume there is no DB updates that are required here?

File: application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/Program.cs (hunk 1 above)
Comment (andrewmorris43) [reply application/CohortManager/src/Functions/CaasIntegration/receiveCaasFile/Program.cs:24]: Updated

File: application/CohortManager/src/Functions/Shared/Data/Database/ScreeningServiceData.cs (hunk 2 above)
Comment (andrewmorris43) [reply application/CohortManager/src/Functions/Shared/Data/Database/ScreeningServiceData.cs]: Done

File: application/CohortManager/src/Functions/Shared/Data/Database/ScreeningServiceData.cs (hunk 3 above)
Comment (andrewmorris43) [reply application/CohortManager/src/Functions/Shared/Data/Database/ScreeningServiceData.cs]: Done

**Comment by andrewmorris43** [comment]
> @andrewmorris43 The regex for checking the file name needs changing to not explicitly reference BSS and maybe change the error message for that exception of that back of that failure

I have updated the regex to remove the hardcoded BSS. It now matches at least 1 word character before the first underscore.

**Review by MWClayson-NHS — APPROVED** [review]
Approved!

**Pull request merged**